File: leetcode/src/main/java/leetcode/array/FirstMissingPositive.java
package leetcode.array;

import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;


/**
 * https://leetcode.com/problems/first-missing-positive/
 *
 * Cho một dãy số nguyên, tìm số nguyên dương nhỏ nhất bị thiếu trong dãy
 *
 * [1, 2, 0]   ->  3
 * [7, 8, 9]   ->  1
 */
public class FirstMissingPositive {

    public static int firstMissingPositive(int[] nums) {
        Set<Integer> visited = new HashSet<>();
        Queue<Integer> queue = new PriorityQueue<>();
        for (int num : nums) {
            if (num > 0 && !visited.contains(num)) {
                queue.add(num);
                visited.add(num);
            }
        }
        int current = 0; // Bắt đầu từ 0
        while (!queue.isEmpty()) {
            // Nếu current + 1 (số kế tiếp) không phải là số kế tiếp (ở đỉnh) thì nó là số bị thiếu
            if (current + 1 != queue.peek()) {
                return current + 1;
            }
            current = queue.poll(); // Gán số ở đỉnh là số hiện tại
        }
        return current + 1;
    }


    @Test
    public void test() {
        assertEquals(3, firstMissingPositive(new int[]{1, 2, 0}));
        assertEquals(2, firstMissingPositive(new int[]{3, 4, -1, 1}));
        assertEquals(1, firstMissingPositive(new int[]{7, 8, 9, 11, 12}));
        assertEquals(3, firstMissingPositive(new int[]{0, 2, 2, 1, 1}));
    }

//    public static int firstMissingPositive(int[] nums) {
//        int current = 0; // Số nguyên dương hiện tại
//        while (!queue.isEmpty()) {
//            if (++current != queue.poll()) {
//                return current;
//            }
//        }
//    }

}

File: py-algorithm/arrays/first_missing_positive.py
import heapq
import unittest
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/first-missing-positive/
    
    Cho mảng số nguyên. Tìm số nguyên còn thiếu đầu tiên nhỏ nhất của mảng số nguyên
"""


class Solution(unittest.TestCase):

    @staticmethod
    def firstMissingPositive(nums: List[int]) -> int:
        # Xóa các phần tử < 0 và trùng nhau
        nums = list(set(filter(lambda t: t > 0, nums)))
        heapq.heapify(nums)
        missing = 0
        while nums:
            # Nếu missing + 1 không phải là số ở đỉnh heap thì nó là số cần tìm
            if missing + 1 != nums[0]:
                return missing + 1

            missing = heapq.heappop(nums)
        # Chạy hết mảng thì phần tử là số lớn nhất (giờ là missing) + 1
        return missing + 1

    def test(self):
        # self.assertEqual(3, self.firstMissingPositive([1, 2, 0]))
        self.assertEqual(2, self.firstMissingPositive([3, 4, -1, 1]))


if __name__ == '__main__':
    unittest.main()
